### PR TITLE
Undo last commit and amend context menu option

### PR DESCRIPTION
**Pull request by kuychaco**
This PR will introduce an "Undo" button that appears only on the most recent commit. Clicking it will do a soft reset to the parent of HEAD, which will restore the staged changes to the state prior to the last commit. The commit message box will be populated with the message of the undone commit.

The amend checkbox will be replaced with a context-menu option that appears only on the most recent commit. Selecting this will run `git commit --amend`. If the commit message box is empty, the previous commit message will be used, otherwise the new commit will have the message from the box. 

It looks pretty much the same, but note the presence of the "undo" button and the absence of the "amend" checkbox.

### Undo
Note that the commit message and co-author fields are repopulated, and changes from the commit appear back on the stage
undo2

TODO:
 - [x] add tests for amend.
    - [x] that staged changes disappear and are incorporated into the most recent commit.
    - [x] if new commit message is provided, use that.  otherwise default to message from previous commit.
     - [x] if no commits exist are we handling that edge case?  (might be covered by another test)
 - [x] expanded commit box manual testing
 - [x] remove checkbox from ui
 - [x] rip out current amend state logic
 - [x] fix any tests that fail as a result of that.
  - [x] get rid of "regular" commit message verbage
  - [x] fix bug amending with coauthors (see gif)
  - [x] handle case when user simply wants to amend message
  - [x] when commiting with expanded editor and co authors, ensure new lines separate co author trailers.
  - [x] make toggle co author icon clickable when commit message is expanded
  - [x] move tests for `addTrailersToCommitMessage` to GSOS tests

Write test cases for `commitIsEnabled` with and without `amend` flag
- staged changes only
- commit message only
- coauthors only
- combinations of the above
- noop if none of the above

### CODE REVIEW

**Comment by smashwilson** [comment]
tenor

**Comment by BinaryMuse** [comment]
Finally! 😉

**Comment by smashwilson** [comment]
Wait. I thought I fixed this one...

```
Git commands for CompositeGitStrategy made of [GitShellOutStrategy] ssh authentication fails the command on dialog cancel - ssh authentication
```

ಠ_ಠ

**Comment by smashwilson** [comment]
... Oh, wait. #1355 hasn't been merged yet! Never mind me.

File: lib/git-shell-out-strategy.js [hunk 1]
```diff
@@ -723,6 +739,16 @@ export default class GitShellOutStrategy {
     return this.exec(args, {useGitPromptServer: true, writeOperation: true});
   }
 
+  /**
+   * Undo Operations
+   */
+  reset(type, revision = 'HEAD') {
+    const validTypes = ['soft'];
```
Comment (smashwilson) [inline lib/git-shell-out-strategy.js]: No `'hard'` or `'mixed'` yet? :smile:

File: lib/controllers/commit-controller.js [hunk 2]
```diff
@@ -229,6 +178,11 @@ export default class CommitController extends React.Component {
     }
   }
 
+  @autobind
+  isCommitMessageEditorExpanded() {
+    return this.getCommitMessageEditors().length > 0;
+  }
```
[the remark this answers is not in the document]
Comment (kuychaco) [reply lib/controllers/commit-controller.js]: yup, good catch 👍

File: lib/git-shell-out-strategy.js [hunk 3]
```diff
@@ -399,20 +398,49 @@ export default class GitShellOutStrategy {
...
-    if (typeof message === 'object') {
-      args = args.concat(['-F', message.filePath]);
-    } else if (typeof message === 'string') {
-      args = args.concat(['-m', message]);
+  async commit(rawMessage, {allowEmpty, amend, coAuthors} = {}) {
+    const args = ['commit', '--cleanup=strip'];
+
+    if (typeof rawMessage !== 'string') {
+      throw new Error(`Invalid message type ${typeof message}. Must be string.`);
+    }
```
[the remark this answers is not in the document]
Comment (kuychaco) [reply lib/git-shell-out-strategy.js]: yup, we can drop this. We used to accept an object that had a filePath key that pointed to the ATOM_COMMIT_EDIT_MSG file and we'd pass that as an option to `git commit`. But now since we're doing some commit message processing to add co-author trailers we're reading the message from the editor prior to committing, rather than the filesystem at commit time. All that to say, this check is less meaningful now that we always expect a string.

File: lib/git-shell-out-strategy.js (hunk 1 above)
Comment (kuychaco) [reply lib/git-shell-out-strategy.js]: heh nope, not yet. @annthurium and I (and maybe @simurai?) talked about it a bit and decided to hold off on those until we have more thorough discussion about UX and other implications. For example, if we implement a way to do a hard reset then we probably want to add some safeguards so that users don't accidentally blow away workdir changes in a way that is unrecoverable. And if someone accidentally does a mixed reset after carefully staging changes, then we've just blown away their index state and we'd probably want to offer a way for them to get back to the state they were in before the reset. 

All of this to say, I'd like to have more developed thoughts on the recovery story if users accidentally do one of these operations. Or alternatively, decide on a UX that makes it really hard for users to accidentally or incorrectly do one of these operations. What do you think @smashwilson 💭

**Comment by kuychaco** [comment]
@smashwilson thanks for the review! I'll go ahead and merge this as-is and incorporate changes based on your comments in #1374 so that we don't have to wait for CI to turn green

**Comment by marcus-blaettermann-ventoro** [comment]
This PR breaks a workflow. Previously it was possible to see the changes in the last commit by clicking the amend checkbox. Also it was possible to see what the new commit would look like before amending.

Now if I want to do that, I have to undo the last commit so it goes to the staging area. But if it turns no amend was necessary the last commit is already destroyed. This especially severe when doing interactive rebase because this could lead to an unnecessary rewrite of the complete history even when not amending anything.

It should at least be possible to undo the undo, so I can get the old commit back.
